src/jit/gentree.cpp

```diff
@@ -12408,6 +12292,171 @@ GenTreePtr Compiler::gtFoldExprSpecial(GenTreePtr tree)
     return op;
 }
 
+//------------------------------------------------------------------------
+// gtTryRemoveBoxUpstreamEffects: given an unused value type box,
+//    try and remove the upstream allocation and unnecessary parts of
+//    the copy.
+//
+// Arguments:
+//    op  -- the box node to optimize
+//
+// Return Value:
+//    True if the op was a value type box and the upstream effects
+//    were removed. Note parts of the copy tree may remain, if the
+//    copy source had side effects.
+//
+//    False if op was not a value type box or the upstream effects
+//    could not be removed.
+//
+// Notes:
+//    Value typed box gets special treatment because it has associated
+//    side effects that can be removed if the box result is not used.
+//
+//    If removal fails, is is possible that a subsequent pass may be
+//    able to optimize.  Blocking side effects may now be minimized
+//    (null or bounds checks might have been removed) or might be
+//    better known (inline return placeholder updated with the actual
+//    return expression). So the box is perhaps best left as is to
+//    help trigger this re-examination.
+
+bool Compiler::gtTryRemoveBoxUpstreamEffects(GenTreePtr op)
+{
+    JITDUMP("gtTryRemoveBoxUpstreamEffects called for [%06u]\n", dspTreeID(op));
+
+    if (!op->IsBoxedValue())
```

Comment:
I'm mildly surprised you're checking `IsBoxedValue` here rather than asserting it here and expecting callers to check.